core/common/scale.cpp

```diff
@@ -0,0 +1,506 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "scale.hpp"
+
+#include <assert.h>
+#include <limits>
+#include <tuple>
+#include <type_traits>
+
+namespace kagome::common {
+
+  namespace scale {
+
+    namespace impl {
+      /**
+       * @brief Union class that helps representing integer value as an array
+       * @tparam T Integer type
+       */
+      template <class T>
+      union AsArray {
+        explicit AsArray(const T &t) : value_{t} {}
+
+        ///
+        uint8_t *begin() {
+          return static_cast<uint8_t *>(array_);
+        }
+
+        uint8_t *end() {
+          return begin() + size();
+        }
+
+        static const size_t size() {
+          return sizeof(T);
+        };
+
+        AsArray reversed() const {
+          AsArray r{value_};
+          for (auto left = r.begin(), right = r.end(); left < right;
+               ++left, --right) {
+            std::swap(*left, *right);
+          }
+          return r;
+        }
+
+        inline auto value() const {
+          return value_;
+        }
+
+        auto toLittleEndian() const {
+          return reverseIfHostNotLittleEndian();
+        }
+
+        auto fromLittleEndian() const {
+          return reverseIfHostNotLittleEndian();
+        }
+
+       private:
+        inline auto reverseIfHostNotLittleEndian() const {
+          if (__BYTE_ORDER == __LITTLE_ENDIAN) {
```

Comment:
Search engines suggest to use `__BYTE_ORDER__` & `__ORDER_LITTLE_ENDIAN__`.

Anyway it's better to use more standard ways, e.g [like unions](https://stackoverflow.com/a/1001373)

Reply:
https://stackoverflow.com/a/38141476/1953079

Reply:
For now, I replaced it by __BYTE_ORDER__ == __LITTLE_ENDIAN__

Reply:
@Warchant me either, but we don't have c++20

The approach above, in fact, can be a constexpr function, thus leading to compile-time evaluation as well

Reply:
Is my choice ok in your opinion?

Reply:
@l4l @masterjedy we do have C++20-compliant compilers - gcc-8